src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs

```diff
@@ -78,6 +78,37 @@ public async Task ApplyProjectBuildAsync(IProjectVersionedValue<IProjectSubscrip
             }
         }
 
+        public Task ApplyProjectUpdatedAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> update, bool isActiveContext, CancellationToken cancellationToken)
```

Comment:
This is very different to how I saw this working, and will likely result in a few behaviors that I don't think are great:

1) Rename used to happen in response to a user action. Now it happens based on any change to the project, whether it was user initiated or not. Actions such as  switching branches, renaming Explorer, etc (if I'm reading correctly) will result in a rename prompt.

2) As we avoid overlapping this method, we'll block updating the language service while we're waiting for the user to confirm and for Roslyn to perform the rename. For example, rename a file -> dialog opens -> user goes has coffee -> comes back and confirms -> the design-time build results as a result of the rename (or anything else that happened in the background) will be blocked until we've finished the rename.

The way I saw this working is like so;

We get called back (like before) based on the user action - either we use existing hooks, or write new ones in CPS if we don't have the right ones.
That component waits until the Solution tree has the updated result of the rename and then waits on the [language service to update to that same version](https://github.com/dotnet/project-system/issues/3425), and then performs the rename.

Reply:
> 1. Rename used to happen in response to a user action. Now it happens based on any change to the project, whether it was user initiated or not. Actions such as  switching branches, renaming Explorer, etc (if I'm reading correctly) will result in a rename prompt.

Unless _I_ have read things incorrectly this case cannot happen. CPS cannot detect a rename via file changes alone. Someone had to modify the `IVsHierarchy` item.  It possible some extension is VS will do this but from our perspective renames are VS initiated actions.  Please correct me if I am wrong and there is some operation I can do on the file-system that causes CPS to believe it is a rename.

> 2. As we avoid overlapping this method, we'll block updating the language service while we're waiting for the user to confirm and for Roslyn to perform the rename. For example, rename a file -> dialog opens -> user goes has coffee -> comes back and confirms -> the design-time build results as a result of the rename (or anything else that happened in the background) will be blocked until we've finished the rename.

Once we acquire the roslyn workspace any changes to that workspace will make us unable to successfully apply a rename operation. If we allow changes to happen under us then the user could `rename a file -> dialog opens -> user confirms -> rename fails as snapshot is out of date`. Because we pop a modal dialog I felt it was appropriate for it to truly be modal and we block other operations while we wait for the users decision.

> We get called back (like before) based on the user action - either we use existing hooks, or write new ones in CPS if we don't have the right ones.
That component waits until the Solution tree has the updated result of the rename and then waits on the [language service to update to that same version](https://github.com/dotnet/project-system/issues/3425), and then performs the rename.

This API would be useful for a lot of scenarios but not for rename. By the time we get called any number of other operations could have happened that invalidate our workspace so we need some way to stop the world in order for the rename to succeed.  What you outlined is essentially how rename works today.  We get notified from CPS that a rename has happened, we register with roslyn to receive workspace changes and wait to observe the rename change.  In practice these events almost never align such that a rename succeeds.  I see no way around this.

Reply:
> Someone had to modify the `IVsHierarchy` item

There's no guaranteed relationship between a rename showing up in the diff and a user initiating a rename. This is an API, and a component, extension, xxx in the background could rename the item and you will prompt in response to it. The other thing that is problematic that we may change the places that we start hinting, for example, if we wanted to make sure that the file stayed open in the the RDT if someone renamed it outside of VS, we'd start adding hints based on the file change rename notifications, and this will start prompting.

_Rename hints/Renames in the project diff  != user initiated rename._

>
By the time we get called any number of other operations could have happened that invalidate our workspace so we need some way to stop the world in order for the rename to succeed

I'm unconvinced by this argument. You are only stopping a single project from updating the workspace and not blocking any other changes. When I rename a type in the editor, this doesn't block us from updating the workspace - so why do we here?

Reply:
> This is an API, and a component, extension, xxx in the background could rename the item and you will prompt in response to it.

That seems fine, or are you saying we need to add something into the Shell that intercepts `IVsHierarchy` renames and tags them somehow so we can track which renames are initiated by VS extensions and which are user initiated? 

> Rename hints/Renames in the project diff != user initiated rename.

I do not know of a way to guarantee that commands issued by VS were only done by a human and not some use of an API.

> I'm unconvinced by this argument. You are only stopping a single project from updating the workspace and not blocking any other changes.

I don't know what else to tell you, this is why rename type to match file does not work today in CPS. It works in legacy today because this is synchronous and requires the UI thread (where code actions are applied) so this situation cannot happen. 

I originally had rename being fire-and-forget with the threading service but ran into issues. I can experiment and see if there is a version that has a high probability of success.

Essentially we want to offer to rename the type as soon as possible.  Users Renames -> CPS Processes Change -> Project System Processes Change -> Roslyn Processes Change -> Rename Offered. It is possible for an operation to happen and invalidate workspace snapshot obviously, but with the current project being the most common place edits are coming from it seemed that pausing for the user gives us the greatest probability of success.

> When I rename a type in the editor, this doesn't block us from updating the workspace - so why do we here?

Because roslyn has a snapshot model and changes made to the workspace using interactive rename are separate from updates from the workspace.  If you did a `git checkout` during interactive rename it would also fail to commit because the soltuion Id would be different. Interactive rename also puts the editor in a mode where is doesn't allow other commands to process or cancels the rename.

Reply:
I believe this is fixed